AppDB/appscale/datastore/fdb/fdb_datastore.py

```diff
@@ -0,0 +1,429 @@
+from appscale.datastore.fdb.gc import GarbageCollector
+from appscale.datastore.fdb.indexes import (
+  get_order_info, IndexManager, KEY_PROP)
+from appscale.datastore.fdb.transactions import TransactionManager
+from appscale.datastore.fdb.utils import (
+  fdb, next_entity_version, DS_ROOT, ScatteredAllocator, TornadoFDB)
+
+sys.path.append(APPSCALE_PYTHON_APPSERVER)
+from google.appengine.datastore import entity_pb
+
+logger = logging.getLogger(__name__)
+
+
+class FDBDatastore(object):
+  """ A datastore implementation that uses FoundationDB. """
+
+  def __init__(self):
+    self.index_manager = None
+    self._data_manager = None
+    self._db = None
+    self._scattered_allocator = ScatteredAllocator()
+    self._tornado_fdb = None
+    self._tx_manager = None
+    self._gc = None
+
+  def start(self):
+    self._db = fdb.open()
+    self._tornado_fdb = TornadoFDB(IOLoop.current())
+    ds_dir = fdb.directory.create_or_open(self._db, DS_ROOT)
+    project_cache = ProjectCache(self._tornado_fdb, ds_dir)
+    self._data_manager = DataManager(self._tornado_fdb, project_cache)
+    self.index_manager = IndexManager(
+      self._db, self._tornado_fdb, self._data_manager, project_cache)
+    self._tx_manager = TransactionManager(self._tornado_fdb, project_cache)
+    self._gc = GarbageCollector(
+      self._db, self._tornado_fdb, self._data_manager, self.index_manager,
+      self._tx_manager, project_cache)
+    self._gc.start()
+
+  @gen.coroutine
+  def dynamic_put(self, project_id, put_request, put_response):
+    #logger.debug('put_request:\n{}'.format(put_request))
+    project_id = decode_str(project_id)
+    # TODO: Enforce max key length (100 elements).
+    # Enforce max element size (1500 bytes).
+    # Enforce max kind size (1500 bytes).
+    # Enforce key name regex (reserved names match __.*__).
+
+    if put_request.auto_id_policy() != put_request.CURRENT:
+      raise BadRequest('Sequential allocator is not implemented')
+
+    tr = self._db.create_transaction()
+
+    if put_request.has_transaction():
+      logger.debug('put in tx: {}'.format(put_request.transaction().handle()))
+      yield self._tx_manager.log_puts(tr, project_id, put_request)
+      writes = [(entity.key(), None, None, None)
+                for entity in put_request.entity_list()]
+    else:
+      futures = []
+      for entity in put_request.entity_list():
+        futures.append(self._upsert(tr, entity))
+
+      writes = yield futures
+
+    old_entities = [(old_entity, old_vs) for _, old_entity, old_vs, _ in writes
+                    if old_entity is not None]
+    vs_future = None
+    if old_entities:
+      vs_future = tr.get_versionstamp()
+
+    yield self._tornado_fdb.commit(tr)
+
+    if old_entities:
+      self._gc.clear_later(old_entities, vs_future.wait().value)
+
+    for key, _, _, new_version in writes:
+      put_response.add_key().CopyFrom(key)
+      if new_version is not None:
+        put_response.add_version(new_version)
```

Comment:
I'm not sure what the behavior is for puts inside a transaction since I don't think they can be given versions yet. I'll investigate. The version field is not populated in the Cassandra backend.

Reply:
After testing on GAE, I can confirm that puts inside a transaction are not given a version. This implementation behaves the same as GAE in that there will either be a version assigned for each entity in the response (if a txid is not given) or there will be no versions assigned.